In [1]:
pip install pyEDFlib

Note: you may need to restart the kernel to use updated packages.


In [2]:
from pyedflib import highlevel
import numpy as np
import glob

path = '/Users/josangmin/Downloads/sleep-edf-database-expanded-1.0.0/sleep-cassette'
data_list = glob.glob(path+'/**')

print(len(data_list))

trains = [x for x in data_list if x.endswith('PSG.edf')]
labels = [x for x in data_list if x.endswith('Hypnogram.edf')]

print('number of train samples:',len(trains))
print('number of labels:',len(labels))

print(highlevel.read_edf(trains[0]))
print('')
print(highlevel.read_edf(labels[0]))

306
number of train samples: 153
number of labels: 153
([array([-20.19047619,  -3.5030525 , -27.52136752, ..., -61.47496947,
       -50.38217338, -22.98778999]), array([ -6.18925519, -10.43589744,  -6.64102564, ...,   1.03907204,
        -3.20757021, -23.08547009]), array([-85.43956044, -25.02930403,  14.94139194, ..., -48.19413919,
        -1.41025641, -14.12820513]), array([256., 255., 266., ..., 269., 308., 266.]), array([3.382, 3.462, 3.486, ..., 3.27 , 3.296, 3.326]), array([14.868, 14.87 , 14.878, ..., 15.029, 15.003, 14.978]), array([964., 968., 967., ..., 814., 847., 833.])], [{'label': 'EEG Fpz-Cz', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 198.0, 'physical_min': -197.0, 'digital_max': 2047, 'digital_min': -2048, 'prefilter': 'HP:0.5Hz LP:100Hz [enhanced cassette BW]', 'transducer': 'Ag-AgCl electrodes'}, {'label': 'EEG Pz-Oz', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 185.0, 'physical_min'

------

153개의 sample이 존재합니다.

highlevel.read_edf return 값이 signals, signal_headers, header 3개입니다.

따로 따로 값을 출력해보겠습니다.

In [3]:
signals = highlevel.read_edf(trains[0])[0]
#signals : np.ndarray or list
#the signals of the chosen channels contained in the EDF.

print(signals)
print(len(signals))
for i in range(len(signals)):
    print(len(signals[i]))


[array([-20.19047619,  -3.5030525 , -27.52136752, ..., -61.47496947,
       -50.38217338, -22.98778999]), array([ -6.18925519, -10.43589744,  -6.64102564, ...,   1.03907204,
        -3.20757021, -23.08547009]), array([-85.43956044, -25.02930403,  14.94139194, ..., -48.19413919,
        -1.41025641, -14.12820513]), array([256., 255., 266., ..., 269., 308., 266.]), array([3.382, 3.462, 3.486, ..., 3.27 , 3.296, 3.326]), array([14.868, 14.87 , 14.878, ..., 15.029, 15.003, 14.978]), array([964., 968., 967., ..., 814., 847., 833.])]
7
8412000
8412000
8412000
84120
84120
84120
84120


signals에는 list 요소가 7개이고, 각 요소에 array(신호)들이 들어있습니다. 

In [7]:
signal_headers = highlevel.read_edf(trains[0])[1]
#signal_headers : list
#one signal header for each channel in the EDF.

print('number of channels:',len(signal_headers))
print('\n')

for i in range(len(signal_headers)):
    print(signal_headers[i])
    print(len(signal_headers[i]))
    print('\n')


number of channels: 7


{'label': 'EEG Fpz-Cz', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 198.0, 'physical_min': -197.0, 'digital_max': 2047, 'digital_min': -2048, 'prefilter': 'HP:0.5Hz LP:100Hz [enhanced cassette BW]', 'transducer': 'Ag-AgCl electrodes'}
10


{'label': 'EEG Pz-Oz', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 185.0, 'physical_min': -185.0, 'digital_max': 2047, 'digital_min': -2048, 'prefilter': 'HP:0.5Hz LP:100Hz [enhanced cassette BW]', 'transducer': 'Ag-AgCl electrodes'}
10


{'label': 'EOG horizontal', 'dimension': 'uV', 'sample_rate': 3000.0, 'sample_frequency': 3000.0, 'physical_max': 927.0, 'physical_min': -933.0, 'digital_max': 2047, 'digital_min': -2048, 'prefilter': 'HP:0.5Hz LP:100Hz [enhanced cassette BW]', 'transducer': 'Ag-AgCl electrodes'}
10


{'label': 'Resp oro-nasal', 'dimension': '', 'sample_rate': 30.0, 'sample_frequency': 30.0, 'physical_max': 2047.0, 'physical_m

signal_headers에는 신호에 대한 정보들이 존재합니다. 

논문과 웹사이트에서는 sample rate가 100Hz, 1Hz라고 나오는데, 3000, 30으로 나옵니다.

이는 epoch 사이즈가 30초이기에, 30초당 3000번, 즉 100Hz라고 이해하면 될것 같습니다. 

In [64]:
header = highlevel.read_edf(trains[0])[2]
#header : dict
#the main header of the EDF file containing meta information.
print(header)
print(len(header))

{'technician': '', 'recording_additional': '', 'patientname': '', 'patient_additional': '', 'patientcode': '', 'equipment': '', 'admincode': '', 'gender': '', 'startdate': datetime.datetime(1989, 5, 8, 16, 18), 'birthdate': '', 'annotations': []}
11


header에는 datetime 말고는 정보가 없습니다. 

trains data에 말고 label_data에 header에 정보가 있습니다.

labels 데이터를 살펴보겠습니다.

In [76]:
signals = highlevel.read_edf(labels[0])[0]
#signals : np.ndarray or list
#the signals of the chosen channels contained in the EDF.
print(signals)

signal_header = highlevel.read_edf(labels[0])[1]
print(signal_header)


[]
[]


signals와 signal_header에 아무것도 없는 것을 확인 할 수 있습니다.

In [8]:
header = highlevel.read_edf(labels[0])[2]
print(len(header))
print(header)


11
{'technician': '', 'recording_additional': '', 'patientname': 'Female 66yr', 'patient_additional': '', 'patientcode': '', 'equipment': '', 'admincode': '', 'gender': 'Female', 'startdate': datetime.datetime(1989, 5, 17, 16, 46), 'birthdate': '', 'annotations': [[0.0, 24390.0, 'Sleep stage W'], [24390.0, 210.0, 'Sleep stage 1'], [24600.0, 30.0, 'Sleep stage 2'], [24630.0, 30.0, 'Sleep stage W'], [24660.0, 30.0, 'Sleep stage 1'], [24690.0, 300.0, 'Sleep stage 2'], [24990.0, 60.0, 'Sleep stage 1'], [25050.0, 180.0, 'Sleep stage 2'], [25230.0, 30.0, 'Sleep stage 1'], [25260.0, 60.0, 'Sleep stage 2'], [25320.0, 60.0, 'Sleep stage 1'], [25380.0, 1140.0, 'Sleep stage 2'], [26520.0, 30.0, 'Sleep stage 3'], [26550.0, 30.0, 'Sleep stage 2'], [26580.0, 30.0, 'Sleep stage 3'], [26610.0, 60.0, 'Sleep stage 2'], [26670.0, 90.0, 'Sleep stage 3'], [26760.0, 120.0, 'Sleep stage 2'], [26880.0, 60.0, 'Sleep stage 3'], [26940.0, 60.0, 'Sleep stage 2'], [27000.0, 90.0, 'Sleep stage 3'], [27090.0, 60.0, 

--------

labels data에 주석이 달려 있습니다. 

[시작 시간, 지속시간, 수면 상태]

이런식으로 labeling되어 있습니다.

label 구성: 
"Sleep stage ?",
"Sleep stage W",
"Sleep stage 1",
"Sleep stage 2",
"Sleep stage 3",
"Sleep stage 4",
"Sleep stage R"


Get information EDF files by other mathod

-----

In [6]:
import pyedflib
edf = pyedflib.EdfReader(trains[1])
print('labels:'+str(edf.getSignalLabels()))
print("Duaration:"+str(edf.getFileDuration()))
print("Freq.:"+str(edf.getSampleFrequencies()))
print("N-Sample(=Freq x Duaration):"+str(edf.getNSamples()))
print("Date:"+str(edf.getStartdatetime()))

labels:['EEG Fpz-Cz', 'EEG Pz-Oz', 'EOG horizontal', 'Resp oro-nasal', 'EMG submental', 'Temp rectal', 'Event marker']
Duaration:82260
Freq.:[3000 3000 3000   30   30   30   30]
N-Sample(=Freq x Duaration):[8226000 8226000 8226000   82260   82260   82260   82260]
Date:1989-07-12 16:09:00


Extract EEG imformation and save to csv files

In [ ]:
import pandas as pd
columns = ['EEG Fpz-Cz', 'EEG Pz-Oz']
for i in range(len(trains)):
    edf = pyedflib.EdfReader(trains[i])
    a = edf.readSignal(0)
    b = edf.readSignal(1)
    data = np.array([a,b])
    data = data.transpose()
    df = pd.DataFrame(data = data, columns = columns)
    c = df_files["patient_id"][i]
    df.to_csv('%i\n.csv' % i)
